src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -0,0 +1,599 @@
+}
+
+model::offset cloud_topic_partition::start_offset() const {
+    // Ask partition for its start offset
+    // TODO: ask dl_stm for its start offset
+    // the 'partition::sync_kafka_start_offset_override' is not invoked here
+    // because it's tied to both archival_metadata_stm and log_eviction_stm.
+    // For cloud topics we will do log eviction differently and the
+    // DeleteRecords API is not implemented yet. So the code is just
+    // using the start_offset of the Raft log at the moment which is incorrect.
+    auto so = _partition->raft_start_offset();
+    auto kso = _partition->get_offset_translator_state()->from_log_offset(so);
+    return kso;
+}
+
+ss::future<result<model::offset, error_code>>
+cloud_topic_partition::sync_effective_start(model::timeout_clock::duration) {
+    // TODO: ask dl_stm
+    co_return start_offset();
+}
+
+model::offset cloud_topic_partition::high_watermark() const {
+    auto ot_state = _partition->get_offset_translator_state();
+    return ot_state->from_log_offset(_partition->high_watermark());
+}
+
+checked<model::offset, error_code>
+cloud_topic_partition::last_stable_offset() const {
+    auto maybe_lso = _partition->last_stable_offset();
+    if (maybe_lso == model::invalid_lso) {
+        return error_code::offset_not_available;
+    }
+    auto ot_state = _partition->get_offset_translator_state();
+    return ot_state->from_log_offset(maybe_lso);
+}
+
+bool cloud_topic_partition::is_leader() const {
+    return _partition->is_leader();
+}
+
+ss::future<std::error_code> cloud_topic_partition::linearizable_barrier() {
+    auto r = co_await _partition->linearizable_barrier();
+    if (r) {
+        co_return raft::errc::success;
+    }
+    co_return r.error();
+}
+
+cluster::partition_probe& cloud_topic_partition::probe() {
+    return _partition->probe();
+}
+
+kafka::leader_epoch cloud_topic_partition::leader_epoch() const {
+    return leader_epoch_from_term(_partition->raft()->confirmed_term());
+}
+
+ss::future<storage::translating_reader> cloud_topic_partition::make_reader(
+  storage::log_reader_config cfg,
+  std::optional<model::timeout_clock::time_point> timeout) {
+    vassert(_ct_api != nullptr, "cloud topics api not initialized");
+
+    auto ot_state = _partition->get_offset_translator_state();
+
+    cfg.start_offset = ot_state->to_log_offset(cfg.start_offset);
+    cfg.max_offset = ot_state->to_log_offset(cfg.max_offset);
+    cfg.translate_offsets = storage::translate_offsets::yes;
+    cfg.type_filter = {model::record_batch_type::dl_placeholder};
+
+    // TODO: add code path that fetches metadata from the dl_stm for L1 read
+    // path
+
+    // This part comprises the control plane query. The partition is a metadata
+    // storage for the cloud topic.
+    auto underlying = co_await _partition->make_reader(cfg, timeout);
+    auto placeholders = co_await model::consume_reader_to_chunked_vector(
+      std::move(underlying), model::no_timeout);
```

Comment:
This is fine for now, long term should we do this incrementally? IE wrap the reader so that each chunk of batches returned separately loads from CT. I guess the local disk should be so much faster than object storage that it should be fine to collect everything from local storage first... 🤔

Reply:
Yes, sorry this was my intent with "fine for now" I need a way of specifying in github what is commentary/dialog vs a blocking comment (better code review tools have this...)

Reply:
AFAICT caching readers is a pure artifact of them being implemented on top of file-based stream readers and the overheads associated with that (pre-fetching, file handles, etc...). For example, the readers in the I/O layer using the page cache do not need to be cached.

So we shouldn't assume that we are going to cache the reader, we should do it only if it makes sense.

Reply:
My hunch is that querying the L0 metadata from inside the reader is complicated. You will have a reader that pulls L0 metadata gradually and based on this metadata it also pulls L0 objects from the cloud storage. It is possible to simplify this if the reader fetches all the metadata that it can possibly need and then fetches the L0 objects gradually. And such reader could be expensive to recreate for every fetch request. Maybe there are better approaches but this is the first thing that comes to mind.